akka-projection-testkit/src/main/scala/akka/projection/testkit/scaladsl/ProjectionTestKit.scala

```diff
@@ -99,19 +99,25 @@ final class ProjectionTestKit private[akka] (testKit: ActorTestKit) {
   }
 
   /**
-   * Run a Projection with an attached `TestSink` allowing
-   * control over the pace the elements flow through the Projection.
+   * Run a Projection with an attached `TestSubscriber.Probe` allowing
+   * control over the pace in which the elements flow through the Projection.
    *
-   * The Projection starts as soon as the first element is requested by the `TestSink`, new elements will be emitted
-   * as requested by the `TestSink`. The Projection won't stop by itself, therefore it's recommended to cancel the
-   * `TestSink` probe to gracefully stop the Projection.
+   * The assertion function receives a `TestSubscriber.Probe` that you can use
+   * request elements.
+   *
+   * The Projection starts as soon as the first element is requested by the `TestSubscriber.Probe`, new elements will be emitted
+   * as requested. The Projection is stopped once the assert function completes.
    *
    * @param projection - the Projection to run
+   * @param assertFunction - a function receiving a `TestSubscriber.Probe[Done]`
    */
-  def runWithTestSink[T](projection: Projection[_]): TestSubscriber.Probe[Done] = {
-    val sinkProbe = TestSink.probe[Done](testKit.system.toClassic)
-    // FIXME handler.stop is not called when running like this
-    projection.mappedSource().runWith(sinkProbe)
+  def runWithTestSink[R](projection: Projection[_])(assertFunction: TestSubscriber.Probe[Done] => R): R = {
+    val sinkProbe = projection.mappedSource().runWith(TestSink.probe[Done](testKit.system.toClassic))
+    try {
+      assertFunction(sinkProbe)
+    } finally {
+      sinkProbe.cancel()
```

Comment:
This seems not be working as expected. There is a Cassandra test that is exercising the handler hooks, but they are not being called.

Reply:
I wasn't able to reproduce this locally.

Reply:
Sorry, I didn't commit the Cassandra test. It was just me play with it locally. 

It turns out that this approach won't work because the hook is added to the Future returned by the sink that that only happens when you 'run' the projection. The `runWithTestSink` is slightly different as it doesn't take the same path as the `run`. It attaches the TestSink to the mappedSource and then we loose the stop hook. 

It's also problematic because each Projection has to re-implement this.